Add name option for the index directive

geographika opened the pull request:
Subject: To allow index hyperlinks to be constructed using a new `:name:` option. 
Feedback on approach welcome, and open to attempting to implement a different approach. 

### Feature or Bugfix
- Feature

### Purpose
The pull request allows the hrefs generated by the [Index Directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#index-generating-markup) to be given a name to be used when creating index links rather than simply an integer id that can change if text on a page is reordered. 

For example:

`.. index:: Test1` will generate `test.html#index-0`

Whereas supplying a new name option such as below:

```
.. index:: single: Test1
   :name: mylink
```

Will generate: `test.html#index-mylink`

If the same `name` option is used multiple times for index entries on the same page a warning similar to below is given:

`index.rst:17: WARNING: Duplicate ID: "index-mylink".`

### Detail
An alternative to simply modifying the href would be to construct a proper `:ref:` node. 
This could be achieved using `from sphinx.util.nodes import make_refnode`?

### Relates
-  #1671

geographika committed:
Add name option for index links

tk0miya commented:
Thank you for proposal. I think its concept is good to me. To realize the feature, we need to discuss how to implement it.

`:name:` option is commonly used in directives of reST. And it generates "hyperlink target" and give a name to it as a reference name. So it should be added into `targetnode['names']` instead. Please don't use it to generate node_id. It is internal ID. So users and developers should not know the value of node_id.
https://docutils.sourceforge.io/docs/ref/rst/directives.html#common-options

geographika committed:
Add option to docs

geographika committed:
Append to tagetnode names list

geographika commented:
@tk0miya - thanks for the feedback. Using the `names` list for the target works well. 
The updated pull request allows the following syntax for an index entry:

```
.. index::
    single: execution; context
    triple: module; search; path
    :name: my-index-link

My section to index
```

And then this index entry can be referred to using a standard ref for example:

```
:ref:`My link <my-index-link>`
```

I looked briefly at trying to automatically add a title to the ref based on the index entry, but this probably introduces more complexity for little benefit. It means any use of a `ref ` without a specific title will return a warning:

```
:ref:`my-index-link`
```

Returns:

`WARNING: undefined label: my-index-link (if the link has no caption the label must precede a section header)`

tk0miya commented on doc/usage/restructuredtext/directives.rst:
```diff
@@ -874,6 +874,19 @@ mainly contained in information units, such as the language reference.
+        .. index:: Python
+           :name: py-index
+
+   .. versionadded:: 2.5
```
Our next target is 3.0. So this should be changed.

geographika replied:
Updated

tk0miya commented on sphinx/domains/index.py:
```diff
@@ -68,12 +69,19 @@ class IndexDirective(SphinxDirective):
-    option_spec = {}  # type: Dict
+    option_spec = {
+        'name': directives.unchanged,
+    }  # type: Dict
```
I guess mypy does not warn this line even if no type hints. Could you try that please?

geographika replied:
I removed the type hint here

tk0miya commented on sphinx/domains/index.py:
```diff
@@ -68,12 +69,19 @@ class IndexDirective(SphinxDirective):
+
+        if 'name' in self.options:
+            targetname = self.options['name']
+            targetnode['names'].append(targetname)
```
@shimizukawa Do you have any opinion for this?

We can also give a node id based on given name like this:
```
if 'name' in self.options:
    targetname = self.options['name']
    targetnode = nodes.target('', '', names=[targetname])
else:
    targetid = 'index-%s' % self.env.new_serialno('index')
    targetnode = nodes.target('', '', ids=[targetid])
```

pros:
* We can use natural node_id generated from name. Therefore HTML anchor would be also natural.

cons:
* ID generation rule will be changed. As a result, some of URLs for indices will be changed.

shimizukawa replied:
For cons, `index-<serialno>` name will be changed by many reasons (new index entry is inserted before existing index entries)  so `index-<serialno>` is not a PERMANENT.

So, +1 for `:name:`  option to get PERMANENT target.

tk0miya replied:
Okay, go with above my code.

tk0miya replied:
@geographika Could you update this please?

geographika replied:
Done - although I used targetid rather than targetname as targetid is required later and I would get a `UnboundLocalError` at `indexnode['entries'].extend(process_index_entry(entry, targetid))`

tk0miya replied:
Please use `targetnode['ids'][0]' instead. `document.note_implicit_target()` generates node_id automatically if not assigned yet. So we can use it after doing it.

tk0miya reviewed (commented):
LGTM

tk0miya commented:
```
.. index::
    single: execution; context
    triple: module; search; path
    :name: my-index-link

My section to index
```

I feel this notation is a bit odd. But it is hard to change the syntax of index directive. So +0 for this.

>I looked briefly at trying to automatically add a title to the ref based on the index entry, but this probably introduces more complexity for little benefit.

I understand your idea. But -0 for it. I think an "index" does not have a title.

tk0miya commented:
Oh, I noticed this does not have testcases. I'll consider how to do it later.

geographika committed:
Review updates

geographika committed:
Add basic syntax test for index :name:

geographika committed:
Add checks for references

geographika commented:
> 
> 
> Oh, I noticed this does not have testcases. I'll consider how to do it later.

@tk0miya - thank you for the review. I've added in a test that should ensure the syntax doesn't break anything, and the two labels are created. I'm happy to change to a different approach if this testing isn't appropriate.

tk0miya commented on tests/test_environment_indexentries.py:
```diff
@@ -126,3 +126,21 @@ def test_create_index_by_key(app):
+@pytest.mark.sphinx('dummy', freshenv=True)
+def test_create_index_with_name(app):
+    text = (".. index:: single: docutils\n   :name: ref1\n"
+            ".. index:: see: Python; interpreter\n   :name: ref2\n")
```
Could you separate these text by each `\n` code? It make this text a bit readable.

geographika replied:
Done

tk0miya commented on tests/test_environment_indexentries.py:
```diff
@@ -126,3 +126,21 @@ def test_create_index_by_key(app):
+    assert index[1] == ('P', [('Python', [[], [('see interpreter', [])], None])])
+
+    # check the reference labels are created correctly
+    labels = app.env.domaindata['std']['anonlabels']
```
Please use this code instead. It makes test code robust because it does not touch inside data structure directly.
```
std = app.env.get_domain('std')
assert std.anonlabels['ref1'] == (...)
```
```

geographika replied:
Done

tk0miya commented on tests/test_environment_indexentries.py:
```diff
@@ -126,3 +126,21 @@ def test_create_index_by_key(app):
+
+
+@pytest.mark.sphinx('dummy', freshenv=True)
+def test_create_index_with_name(app):
```
Could you move this above `test_create_index_by_key()`? It would be better to place testcases for index directive nearly.

geographika replied:
Done

tk0miya reviewed (commented):
Thank you for update. Just a little more :-)

tk0miya commented:
I think testcodes are fine with nits.

geographika committed:
Implement review changes

geographika committed:
Review updates

geographika commented:
@tk0miya - thanks for the feedback. I've implemented all the suggestions, and updated the test tp check for both a named ref and an autogenerated index entry.

geographika committed:
Use targetnode['ids']

geographika committed:
Fix whitespace

tk0miya reviewed (approved):
Thank you for update! LGTM!!

tk0miya commented:
Merged 🎉 . Thank you for your effort :-)

geographika commented:
@tk0miya  - many thanks for taking the time to review and walk me through the process
